File: core/src/main/java/org/runetranscriber/core/viking/AngloSaxonRune.java
package org.runetranscriber.core.viking;

import org.runetranscriber.core.DefaultRune;
import org.runetranscriber.core.Rune;
import org.runetranscriber.core.RuneList;

/**
 * Provides an enumeration of runes in the Anglo-Saxon runic alphabet.
 */
public final class AngloSaxonRune implements Rune
{
    /** Rune. */
    public static final AngloSaxonRune FEOH = new AngloSaxonRune("FEOH", "wealth");

    /** Rune. */
    public static final AngloSaxonRune UR = new AngloSaxonRune("UR", "aurochs");

    /** Rune. */
    public static final AngloSaxonRune THORN = new AngloSaxonRune("THORN", "thorn");

    /** Rune. */
    public static final AngloSaxonRune OS = new AngloSaxonRune("OS", "[a] god");

    /** Rune. */
    public static final AngloSaxonRune RAD = new AngloSaxonRune("RAD", "ride");

    /** Rune. */
    public static final AngloSaxonRune CEN = new AngloSaxonRune("CEN", "torch");

    /** Rune. */
    public static final AngloSaxonRune GYFU = new AngloSaxonRune("GYFU", "gift");

    /** Rune. */
    public static final AngloSaxonRune WYNN = new AngloSaxonRune("WYNN", "joy");

    /** Rune. */
    public static final AngloSaxonRune HAEGL = new AngloSaxonRune("HAEGL", "hail (precipitation)");

    /** Rune. */
    public static final AngloSaxonRune NYD = new AngloSaxonRune("NYD", "need, distress");

    /** Rune. */
    public static final AngloSaxonRune IS = new AngloSaxonRune("IS", "ice");

    /** Rune. */
    public static final AngloSaxonRune GER = new AngloSaxonRune("GER", "year, harvest");

    /** Rune. */
    public static final AngloSaxonRune EOH = new AngloSaxonRune("EOH", "yew");

    /** Rune. */
    public static final AngloSaxonRune PEORD = new AngloSaxonRune("PEORD", "(Unknown)");

    /** Rune. */
    public static final AngloSaxonRune EOLH = new AngloSaxonRune("EOLH", "elk-sedge");

    /** Rune. */
    public static final AngloSaxonRune SIGEL = new AngloSaxonRune("SIGEL", "Sun");

    /** Rune. */
    public static final AngloSaxonRune TIW = new AngloSaxonRune("TIW", "Tiw");

    /** Rune. */
    public static final AngloSaxonRune BEORC = new AngloSaxonRune("BEORC", "birch");

    /** Rune. */
    public static final AngloSaxonRune EH = new AngloSaxonRune("EH", "horse");

    /** Rune. */
    public static final AngloSaxonRune MANN = new AngloSaxonRune("MANN", "man");

    /** 21 */
    public static final AngloSaxonRune LAGU = new AngloSaxonRune("LAGU", "lake");

    /** 22 */
    public static final AngloSaxonRune ING = new AngloSaxonRune("ING", "Ing (a hero)");

    /** 23 */
    public static final AngloSaxonRune EOEL = new AngloSaxonRune("EOEL", "estate");

    /** 24 */
    public static final AngloSaxonRune DAEG = new AngloSaxonRune("DAEG", "day");

    /** 25 */
    public static final AngloSaxonRune AC = new AngloSaxonRune("AC", "oak");

    /** Rune. */
    public static final AngloSaxonRune AESC = new AngloSaxonRune("AESC", "ash-tree");

    /** Rune. */
    public static final AngloSaxonRune YR = new AngloSaxonRune("YR", "bow");

    /** Rune. */
    public static final AngloSaxonRune IOR = new AngloSaxonRune("IOR", "eel");

    /** Rune. */
    public static final AngloSaxonRune EAR = new AngloSaxonRune("EAR", "grave");

    /** Punctuation. */
    /** Rune. */
    public static final AngloSaxonRune SPACE = new AngloSaxonRune("SPACE", "space");

    /** Rune. */
    public static final AngloSaxonRune COMMA = new AngloSaxonRune("COMMA", "comma");

    /** Rune. */
    public static final AngloSaxonRune PERIOD = new AngloSaxonRune("PERIOD", "period");

    /** Rune. */
    public static final AngloSaxonRune NEWLINE = new AngloSaxonRune("NEWLINE", "newline");

    /** Ordinal count. */
    private static int ordinalCount = 0;

    /** Values. */
    private static RuneList<AngloSaxonRune> VALUES;

    /**
     * @param number Rune number.
     * 
     * @return rune.
     */
    public final static AngloSaxonRune valueOfNumber(final int number)
    {
        AngloSaxonRune answer = null;

        final RuneList<AngloSaxonRune> values = values();

        if ((0 < number) && (number <= values.size()))
        {
            answer = values.get(number - 1);
        }

        return answer;
    }

    /**
     * @return a copy of the list of rune values.
     */
    public static final RuneList<AngloSaxonRune> values()
    {
        return new RuneList<AngloSaxonRune>(VALUES);
    }

    /** Delegate. */
    private final DefaultRune delegate;

    /**
     * Construct this object.
     * 
     * @param name Name.
     * @param nameMeaning Name meaning.
     */
    private AngloSaxonRune(final String name, final String nameMeaning)
    {
        final String displayName = name.toLowerCase();
        delegate = new DefaultRune(ordinalCount, name, displayName, nameMeaning);

        if (VALUES == null)
        {
            VALUES = new RuneList<AngloSaxonRune>();
        }

        VALUES.add(this);
        ordinalCount++;
    }

    @Override
    public String displayName()
    {
        return delegate.displayName();
    }

    @Override
    public boolean equals(final Object object)
    {
        boolean answer = false;

        if (object == this)
        {
            answer = true;
        }
        else if (object == null)
        {
            answer = false;
        }
        else if (getClass() != object.getClass())
        {
            answer = false;
        }
        else
        {
            final AngloSaxonRune another = (AngloSaxonRune)object;

            answer = delegate.equals(another.delegate);
        }

        return answer;
    }

    @Override
    public int hashCode()
    {
        return delegate.hashCode();
    }

    @Override
    public String name()
    {
        return delegate.name();
    }

    @Override
    public String nameMeaning()
    {
        return delegate.nameMeaning();
    }

    @Override
    public int ordinal()
    {
        return delegate.ordinal();
    }

    @Override
    public String toDebugString()
    {
        return delegate.toDebugString();
    }

    @Override
    public String toString()
    {
        return delegate.toString();
    }
}

File: webapp/src/main/js/AngloSaxonRune.js
/*
 * Provides an enumeration of runes in the Anglo-Saxon runic alphabet.
 */
define(function()
{
    "use strict";
    var AngloSaxonRune =
    {
        FEOH: "feoh",
        UR: "ur",
        THORN: "thorn",
        OS: "os",
        RAD: "rad",
        CEN: "cen",
        GYFU: "gyfu",
        WYNN: "wynn",
        HAEGL: "haegl",
        NYD: "nyd",
        IS: "is",
        GER: "ger",
        EOH: "eoh",
        PEORD: "peord",
        EOLH: "eolh",
        SIGEL: "sigel",
        TIW: "tiw",
        BEORC: "beorc",
        EH: "eh",
        MANN: "mann",
        LAGU: "lagu",
        ING: "ing",
        EOEL: "eoel",
        DAEG: "daeg",
        AC: "ac",
        AESC: "aesc",
        YR: "yr",
        IOR: "ior",
        EAR: "ear",
        SPACE: "space",
        COMMA: "comma",
        PERIOD: "period",
        NEWLINE: "newline",

        properties:
        {
            "feoh":
            {
                name: "FEOH",
                displayName: "feoh",
                fontLetter: "\u16a0",
                phoneme: "f",
            },
            "ur":
            {
                name: "UR",
                displayName: "ur",
                fontLetter: "\u16a2",
                phoneme: "u",
            },
            "thorn":
            {
                name: "THORN",
                displayName: "thorn",
                fontLetter: "\u16a6",
                phoneme: "th",
            },
            "os":
            {
                name: "OS",
                displayName: "os",
                fontLetter: "\u16a9",
                phoneme: "o",
            },
            "rad":
            {
                name: "RAD",
                displayName: "rad",
                fontLetter: "\u16b1",
                phoneme: "r",
            },
            "cen":
            {
                name: "CEN",
                displayName: "cen",
                fontLetter: "\u16b3",
                phoneme: "k",
            },
            "gyfu":
            {
                name: "GYFU",
                displayName: "gyfu",
                fontLetter: "\u16b7",
                phoneme: "g",
            },
            "wynn":
            {
                name: "WYNN",
                displayName: "wynn",
                fontLetter: "\u16b9",
                phoneme: "w",
            },
            "haegl":
            {
                name: "HAEGL",
                displayName: "haegl",
                fontLetter: "\u16bb",
                phoneme: "h",
            },
            "nyd":
            {
                name: "NYD",
                displayName: "nyd",
                fontLetter: "\u16be",
                phoneme: "n",
            },
            "is":
            {
                name: "IS",
                displayName: "is",
                fontLetter: "\u16c1",
                phoneme: "i",
            },
            "ger":
            {
                name: "GER",
                displayName: "ger",
                fontLetter: "\u16c4",
                phoneme: "j",
            },
            "eoh":
            {
                name: "EOH",
                displayName: "eoh",
                fontLetter: "\u16c7",
                phoneme: "eo",
            },
            "peord":
            {
                name: "PEORD",
                displayName: "peord",
                fontLetter: "\u16c8",
                phoneme: "p",
            },
            "eolh":
            {
                name: "EOLH",
                displayName: "eolh",
                fontLetter: "\u16c9",
                phoneme: "x",
            },
            "sigel":
            {
                name: "SIGEL",
                displayName: "sigel",
                fontLetter: "\u16cb",
                phoneme: "s",
            },
            "tiw":
            {
                name: "TIW",
                displayName: "tiw",
                fontLetter: "\u16cf",
                phoneme: "t",
            },
            "beorc":
            {
                name: "BEORC",
                displayName: "beorc",
                fontLetter: "\u16d2",
                phoneme: "b",
            },
            "eh":
            {
                name: "EH",
                displayName: "eh",
                fontLetter: "\u16d6",
                phoneme: "e",
            },
            "mann":
            {
                name: "MANN",
                displayName: "mann",
                fontLetter: "\u16d7",
                phoneme: "m",
            },
            "lagu":
            {
                name: "LAGU",
                displayName: "lagu",
                fontLetter: "\u16da",
                phoneme: "l",
            },
            "ing":
            {
                name: "ING",
                displayName: "ing",
                fontLetter: "\u16dd",
                phoneme: "ng",
            },
            "eoel":
            {
                name: "EOEL",
                displayName: "eoel",
                fontLetter: "\u16df",
                phoneme: "oe",
            },
            "daeg":
            {
                name: "DAEG",
                displayName: "daeg",
                fontLetter: "\u16de",
                phoneme: "d",
            },
            "ac":
            {
                name: "AC",
                displayName: "ac",
                fontLetter: "\u16aa",
                phoneme: "a",
            },
            "aesc":
            {
                name: "AESC",
                displayName: "aesc",
                fontLetter: "\u16ab",
                phoneme: "ae",
            },
            "yr":
            {
                name: "YR",
                displayName: "yr",
                fontLetter: "\u16a3",
                phoneme: "y",
            },
            "ior":
            {
                name: "IOR",
                displayName: "ior",
                fontLetter: "\u16bc",
                phoneme: "ia",
            },
            "ear":
            {
                name: "EAR",
                displayName: "ear",
                fontLetter: "\u16e0",
                phoneme: "ea",
            },
            "space":
            {
                name: "SPACE",
                displayName: "space",
                fontLetter: "\u16eb",
                phoneme: " ",
            },
            "comma":
            {
                name: "COMMA",
                displayName: "comma",
                fontLetter: "\u16ec",
                phoneme: ",",
            },
            "period":
            {
                name: "PERIOD",
                displayName: "period",
                fontLetter: "\u1392",
                phoneme: ".",
            },
            "newline":
            {
                name: "NEWLINE",
                displayName: "newline",
                fontLetter: "\n",
                phoneme: "\n",
            },
        },

        values: [ "feoh", "ur", "thorn", "os", "rad", "cen", "gyfu", "wynn", "haegl", "nyd", "is", "ger", "eoh",
                "peord", "eolh", "sigel", "tiw", "beorc", "eh", "mann", "lagu", "ing", "eoel", "daeg", "ac", "aesc",
                "yr", "ior", "ear", "space", "comma", "period", "newline", ],

        isPunctuation: function(rune)
        {
            return [ AngloSaxonRune.SPACE, AngloSaxonRune.COMMA, AngloSaxonRune.PERIOD, AngloSaxonRune.NEWLINE ]
                    .includes(rune) ||
                    [ " ", ",", ".", "\n" ].includes(rune);
        }
    };

    if (Object.freeze)
    {
        Object.freeze(AngloSaxonRune);
    }

    return AngloSaxonRune;
});
